smbclient/shutil.py

```diff
@@ -0,0 +1,306 @@
+
+
+def rmtree(path, ignore_errors=False, onerror=None, **kwargs):
+    """Recursively delete a directory tree.
+
+    If ignore_errors is set, errors are ignored; otherwise, if onerror
+    is set, it is called to handle the error with arguments (func,
+    path, exc_info) where func is smbclient.listdir, smbclient.remove,
+    or smbclient.rmdir;
+    path is the argument to that function that caused it to fail; and
+    exc_info is a tuple returned by sys.exc_info().  If ignore_errors
+    is false and onerror is None, an exception is raised.
+
+    :param path: The path to remove.
+    :param ignore_errors: The ignore errors flag.
+    :param onerror: The callback executed on errors
+    :param kwargs: Common arguments used to build the SMB Session.
+    :return: True if path is a dir or points to a dir.
+    """
+    if ignore_errors:
+        def onerror(*args):
+            pass
+    elif onerror is None:
+        def onerror(*args):
+            raise
+    try:
+        if islink(path, **kwargs):
+            # symlinks to directories are forbidden, see bug #1669
+            raise OSError("Cannot call rmtree on a symbolic link")
+    except OSError:
+        onerror(islink, path, sys.exc_info())
+        # can't continue even if onerror hook returns
+        return
+
+    names = []
+    try:
+        names = listdir(path, **kwargs)
+    except os_error:
+        onerror(listdir, path, sys.exc_info())
+    for name in names:
+        fullname = join(path, name)
+        try:
+            mode = lstat(fullname, **kwargs).st_mode
+        except os_error:
+            mode = 0
+        if py_stat.S_ISDIR(mode):
+            rmtree(fullname, ignore_errors, onerror, **kwargs)
+        else:
+            try:
+                remove(fullname, **kwargs)
+            except os_error:
+                onerror(remove, fullname, sys.exc_info())
+    try:
+        rmdir(path, **kwargs)
+    except os_error:
+        onerror(rmdir, path, sys.exc_info())
+
+
+def copy2(src, dst, server_side_copy=False, **kwargs):
+    """Copy data and metadata. Return the file's destination.
+
+    Metadata is copied with copystat(). Please see the copystat function
+    for more information.
+
+    The destination may be a directory.
+
+    """
+    if isdir(dst, **kwargs):
+        dst = join(dst, basename(src))
+    if server_side_copy:
+        copyfile_server_side(src, dst, **kwargs)
+    else:
+        copyfile(src, dst, **kwargs)
+    copystat(src, dst, **kwargs)
+
+
+def copystat(src, dst, **kwargs):
+    """Copy file metadata
+
+    Copy the permission bits, last access time, last modification time, and
+    flags from `src` to `dst`. The file contents, owner, and group are
+    unaffected. `src` and `dst` are path names given as strings.
+    """
+    st = stat(src, **kwargs)
+    utime(dst, ns=(st.st_atime_ns, st.st_mtime_ns), **kwargs)
+
+    # TODO: in principle, we can copy the readonly flag via chmod
```

Comment:
This is partly why I don't think we should ship with the copy2 functionality for now. Better to focus on getting a basic file copy that supports all the permutations of local/remote for now.

Reply:
removed